htcanalyze/htcanalyze.py

```diff
@@ -72,34 +72,35 @@ def manage_thresholds(self, resources: dict) -> dict:
         """
         # change to list, to avoid numpy type errors
         resources.update(Usage=list(resources["Usage"]))
+        # TODO: What does i do? Making resources['Requested'] a dict would be more descriptive
         for i, _ in enumerate(resources['Resources']):
             # thresholds used vs. requested
             if float(resources['Requested'][i]) != 0:
 
                 deviation = float(resources['Usage'][i]) / float(
                     resources['Requested'][i])
-
-                # color red
-                if deviation >= 1 + self.bad_usage \
-                        or deviation <= 1 - self.bad_usage:
-                    resources['Usage'][i] = f"[red]" \
-                        f"{resources['Usage'][i]}[/red]"
-                # color yellow
-                elif deviation >= 1 + self.tolerated_usage \
-                        or deviation <= 1 - self.tolerated_usage:
-                    resources['Usage'][i] = f"[yellow]" \
-                        f"{resources['Usage'][i]}[/yellow]"
-                # if nan, mark yellow
-                elif str(resources['Usage'][i]) == "nan":
-                    resources['Usage'][i] = f"[yellow2]" \
-                        f"{resources['Usage'][i]}[/yellow2]"
-                # else it's okay, color green
-                else:
-                    resources['Usage'][i] = f"[green]" \
-                        f"{resources['Usage'][i]}[/green]"
-
+                # TODO: to make it more readable, use a function
+                line_color = get_color(deviation, resources['Usage'][i])
+                resources['Usage'][i] = f"[{line_color}]" \
+                                        f"{resources['Usage'][i]}[/{line_color}]"
         return resources
 
+    def get_color(self, deviation, usage):
```

Comment:
I wouldn't make `get_color` a function, because this contains logic that is only relevant to the code section you pulled it from. The entire idea of `deviation`, `bad_usage`, `tolerated_usage` in combination with colors only applies to this section. If I were to read this function out of context, I would have no clue what any of this means.

However, if colors in other sections of the script depend on similar settings, then it makes sense to make a function that indicates the "level" of the line (i.e., warning level is yellow, error level is red, normal level is green), because that will ensure consistent color usage throughout the code. That would mean, in this case, that there should either be a function or a part in the code above that converts deviation or usage to a warning level, and then submit this warning level to the `get_color` function. 

I furthermore suggest changing the `if` statements to check whether a value is in between two numbers, rather than doing so with `if or`. 

For example:

```
if 1 - self.bad_usage <= deviation <= 1 + self.bad_usage:
    level = 'error'
elif 1 - self.tolerated_usage <= deviation <= 1 + self.tolerated_usage:
    level = 'warning'
elif str(usage) == 'nan':
    level = 'light_warning'  # or some other name to indicate usage of 'yellow2'
else:
    level = 'normal'

color = get_color(level)


def get_color(level: str) -> str:
    colors = {'error': 'red', 'warning': 'yellow', 'light_warning': 'yellow2', 'normal': 'green'}
    return colors[level]
```

Want to change the colors? You only have to do so in `get_color` now. The logic for deciding what is a warning, an error, or normal, now resides were it logically should.

Reply:
This is a good approach, thanks for mentioning, I've implemented this in the newer commits